### PR TITLE
Add abort on warning threshold

### PR DESCRIPTION
**Pull request by lucafedeli88**
This PR adds a feature suggested by @ax3l : the option to abort as soon as a warning above a given priority threshold is raised.
We can now add lines like 
```
warpx.abort_on_warning_threshold = "medium"
```
to abort as soon as a medium priority or a high priority warning is set.
The default is to never abort upon a warning.

I've updated all the tests so that this option is set (if possible) to ``low`` . In some cases I had to set a higher threshold, for instance when the built-in low-resolution lookup tables are used (a ``medium`` priority warning is raised).

I have also updated the documentation.

### CODE REVIEW

**Comment by lucafedeli88** [comment]
Few tests have failed. I still have to understand why....

File: Source/WarpX.cpp [hunk 1]
```diff
@@ -492,8 +508,26 @@ WarpX::ReadParameters ()
...
+            pp_warpx.query("abort_on_warning_threshold", str_abort_on_warning_threshold)){
+            if (str_abort_on_warning_threshold == "high")
+                m_abort_on_warning_threshold = WarnPriority::high;
+            else if (str_abort_on_warning_threshold == "medium" )
+                m_abort_on_warning_threshold = WarnPriority::medium;
+            else if (str_abort_on_warning_threshold == "low")
+                m_abort_on_warning_threshold = WarnPriority::low;
+            else {
+                const auto t_str = str_abort_on_warning_threshold +
+                    "is not a valid option for warpx.abort_on_warning_threshold";
```
Comment (ax3l) [inline Source/WarpX.cpp]:
```suggestion
                    "is not a valid option for warpx.abort_on_warning_threshold (use: low, medium or high)";
```

File: Source/WarpX.H [hunk 2]
```diff
@@ -1005,6 +1006,8 @@ private:
     std::unique_ptr<Utils::WarnManager> m_p_warn_manager;
     // Flag to control if WarpX has to emit a warning message as soon as a warning is recorded
     bool m_always_warn_immediately = false;
+    // Threshold to abort immediately on a warning message
+    std::optional<WarnPriority> m_abort_on_warning_threshold = std::nullopt;
```
Comment (ax3l) [inline Source/WarpX.H:1018]: Nice :D

**Review by ax3l — COMMENTED** [review]
For inputs, we could also add this generally here:
```
https://github.com/ECP-WarpX/WarpX/blob/73858579e90701b5f1fc1ef9009a8067294aa975/Regression/prepare_file_ci.py#L72-L76
```

note: some tests already raise warnings.

**Review by EZoni — COMMENTED** [review]
Thank you for this PR, Luca! The changes look good to me. However, I think I don't understand the OpenMP-related changes I see in Source/Particles/Gather/FieldGather.H. Were they maybe left over from local work on some other branch? Is that maybe why some CI tests are failing?

File: Source/WarpX.cpp (hunk 1 above)
Comment (lucafedeli88) [reply Source/WarpX.cpp]: thanks!

**Comment by lucafedeli88** [comment]
> Thank you for this PR, Luca! The changes look good to me. However, I think I don't understand the OpenMP-related changes I see in Source/Particles/Gather/FieldGather.H. Were they maybe left over from local work on some other branch? Is that maybe why some CI tests are failing?

Ah! Thanks a lot! Yes, I should have removed them... I'll do that immediately...

**Comment by lucafedeli88** [comment]
> > Thank you for this PR, Luca! The changes look good to me. However, I think I don't understand the OpenMP-related changes I see in Source/Particles/Gather/FieldGather.H. Were they maybe left over from local work on some other branch? Is that maybe why some CI tests are failing?
> 
> Ah! Thanks a lot! Yes, I should have removed them... I'll do that immediately...

Done!

**Comment by lucafedeli88** [comment]
> For inputs, we could also add this generally here:
> 
> ```
> https://github.com/ECP-WarpX/WarpX/blob/73858579e90701b5f1fc1ef9009a8067294aa975/Regression/prepare_file_ci.py#L72-L76
> ```
> 
> note: some tests already raise warnings.

The issue with this solution is that (very) few tests actually need the threshold to be set to ``medium`` or even ``high`` . Is it possible to do so while having a default for all the other tests?

**Comment by lucafedeli88** [comment]
I am not sure about what to do with some tests.

For instance,:

- ``/Tests/embedded_circle/inputs_2d`` raises a ``medium`` priority warning due to the lines 
```
ar_ions.species_type = argon
ar_ions.charge = q_e
```
however, this is intentional since in this test we want partially ionized Argon atoms.

- ``Tests/ElectrostaticSphereEB/inputs_3d`` and ``Tests/ElectrostaticDirichletBC/inputs_2d`` raise a ``low`` priority warning stating that the max norm of ``rho`` is 0. I guess that this is OK, since there are no particles, right?

- ``Tests/ElectrostaticSphereEB/inputs_rz`` , ``ElectrostaticSphere/inputs_rz`` , ``Langmuir/inputs_2d_multi_rz_rt`` , ``Tests/Langmuir/inputs_2d_multi_rz_rt`` , ``Tests/Langmuir/inputs_2d_multi_rz_rt ``  , `` Tests/collision/inputs_rz ``, `` Tests/galilean/inputs_rz `` , `` Tests/multi_J/inputs_rz
 `` , `` Tests/scraping/inputs_rz ``,   and `` Tests/SilverMueller/inputs_rz_z `` all raise a mysterious ``low`` priority warning stating that  ``geometry.coord_sys`` is ignored, although  ``geometry.coord_sys`` does not appear in the inputfile!!

- ``Physics_applications/laser_acceleration/inputs_rz`` raises a ``medium`` warning stating that Filter currently not working with FDTD solver in RZ geometry. Should we set ``warpx.use_filter = 0`` in the inputfile ?

- ``Tests/initial_distribution/inputs`` raises a ``high`` priority warning, stating that there is too little work: 1 box for 2 MPI tasks. We should probably disable MPI for this test...

- ``Physics_applications/capacitive_discharge/inputs_2d`` raises a ``medium`` priority warning due to the lines 
```
he_ions.species_type = helium
he_ions.charge = q_e
```
However, it seems to me that we want partially ionized helium in this test

**Comment by EZoni** [comment]
@lucafedeli88 Regarding what you say here about the RZ tests,
> [...] all raise a mysterious low priority warning stating that geometry.coord_sys is ignored, although geometry.coord_sys does not appear in the inputfile!!

I suspect that these lines in `CheckGriddingForRZSpectral` are the ones causing the issue:
https://github.com/ECP-WarpX/WarpX/blob/4e9d98c528dfebf05aa9d07edf451eff29f37f65/Source/Utils/WarpXUtil.cpp#L550-L553

I will open a small PR to remove these lines and see if that works fine without breaking anything else. Will keep you posted.

**Comment by EZoni** [comment]
The issues related to `coord_sys` with the RZ tests and the issue related to MPI decomposition with `initial_distribution` should now be fixed by #2759 and #2760, provided that this PR is rebased on `development`.

Regarding the filter warning for the tests associated with the input file Physics_applications/laser_acceleration/inputs_rz,

> Should we set warpx.use_filter = 0 in the input file?

I think this is related to the open issue #1943. Even though the behavior is not as expected, it's possible that we still want to cover that RZ filtering machinery in CI, just to make sure that it doesn't break (e.g., crashes), on top of not behaving correctly. I'm not sure, though. @dpgrote Dave, what do you think about this?

**Comment by dpgrote** [comment]
Regarding Physics_applications/laser_acceleration/inputs_rz, I would suggest setting `warpx.use_filter = 0`, though this will require resetting the benchmark. This case must have slipped through when the warning was added. @RemiLehe do you have any comment on this?

**Comment by ax3l** [comment]
Oh yes, that is exactly #1943, isn't it?
Update: oh, Edoardo already linked it.

**Comment by EZoni** [comment]
Regarding the discussion above, I'm a little concerned about setting `warpx.use_filter = 0` for the RZ tests where this is showing up. Aren't we worried that, if the combination RZ + FDTD + filter is not used at all in CI, it might break down at some point (meaning that, on top of not yielding physically correct results, it might also start crashing silently, without us noticing through CI)?

**Comment by lucafedeli88** [comment]
https://github.com/ECP-WarpX/WarpX/pull/2765 should fix the ``high`` priority warning raised by ``Tests/initial_distribution/inputs``

**Comment by lucafedeli88** [comment]
> Regarding the discussion above, I'm a little concerned about setting `warpx.use_filter = 0` for the RZ tests where this is showing up. Aren't we worried that, if the combination RZ + FDTD + filter is not used at all in CI, it might break down at some point (meaning that, on top of not yielding physically correct results, it might also start crashing silently, without us noticing through CI)?

If we don't want to support this combination we could maybe add an Abort. Otherwise, we can simply set the threshold to ``high`` for that test and add a short comment.

**Comment by lucafedeli88** [comment]
https://github.com/ECP-WarpX/WarpX/pull/2771 should fix several warnings that we are still getting

**Comment by EZoni** [comment]
> If we don't want to support this combination we could maybe add an Abort. Otherwise, we can simply set the threshold to `high` for that test and add a short comment.

@dpgrote @RemiLehe Do you have preferences regarding @lucafedeli88's question here about filtering with FDTD in RZ geometry? Maybe we should indeed abort the code until the corresponding issue #1943 is fixed/understood, since the code would not return correct results anyways.

**Comment by dpgrote** [comment]
Since the code is not producing correct results with RZ + FDTD + filtering, I would think that this should produce an abort.

**Comment by lucafedeli88** [comment]
I've raised the threshold for the few tests where the charge is redefined for ionizable species.
I've also raised the threshold for tests with the electrostatic solver where max rho was zero.
~I've also set `warpx.self_fields_absolute_tolerance = 1e-06` explicitly (it is the default value) in order to avoid
a low priority warning.~

**Review by EZoni — APPROVED** [review]
Thank you, Luca! Looks good to me.

File: Tools/PerformanceTests/automated_test_6_output_2ppc [hunk 3]
```diff
@@ -62,3 +62,7 @@ diagnostics.diags_names = diag1
 diag1.intervals = 1
 diag1.file_prefix = "./diags/plt"
 diag1.diag_type = Full
+
+# These options are mainly intended for debug purposes
+warpx.abort_on_warning_threshold = "low"
+warpx.always_warn_immediately = 1
```
Comment (ax3l) [inline Tools/PerformanceTests/automated_test_6_output_2ppc]: Is this adding the same two lines to each test?

I would rather like to add them like to add them _in one line_ in `Regression/WarpX-tests.ini` instead:
https://github.com/ECP-WarpX/WarpX/blob/development/Regression/WarpX-tests.ini#L67
```
runtime_params = ...
```

An individual test could then still overwrite this via their own `runtime_params = ...`.

or via
https://github.com/ECP-WarpX/WarpX/blob/53f590c804ea39f040bcbecb5ad7bb6ab939e06e/Regression/prepare_file_ci.py#L72-L76

The rational behind this: it is unnecessarily verbose and we want to add this permanently (like `-Werror`), so that new tests cannot forget to add this line.

Comment (lucafedeli88) [reply Tools/PerformanceTests/automated_test_6_output_2ppc]: I agree with you... I'll do that now

Comment (lucafedeli88) [reply Tools/PerformanceTests/automated_test_6_output_2ppc]: let's see if all the tests pass...

Comment (lucafedeli88) [reply Tools/PerformanceTests/automated_test_6_output_2ppc]: It should be OK now!

**Review by ax3l — APPROVED** [review]
Excellent, thank you! :+1:
